Nuvoton: Adhere to reworked ticker spec to release with Mbed OS 5.9

ccli8 opened the pull request:
### Description

This PR tries to adhere to reworked ticker spec to release with Mbed OS 5.9.

#### Targets
- NUMAKER_PFM_NUC472
- NUMAKER_PFM_M453
- NUMAKER_PFM_M487
- NUMAKER_PFM_NANO130

### Pull request type

    [ ] Fix
    [X] Refactor
    [ ] New target
    [ ] Feature
    [ ] Breaking change

ccli8 committed:
[Nuvoton] Meet new us_ticker HAL spec (Mbed OS 5.9)

1. Add USTICKER in device_has option of targets.json file.
2. Disable ticker interrupt in us_ticker_init
3. Add us_ticker_free
4. Enable interrupt in us_ticker_set_interrupt

ccli8 committed:
[Nuvoton] Meet new lp_ticker HAL spec (Mbed OS 5.9)

1. Add LPTICKER in device_has option of targets.json file.
2. Disable ticker interrupt in lp_ticker_init
3. Add lp_ticker_free
4. Enable interrupt in lp_ticker_set_interrupt

ccli8 committed:
Fix Greentea test code

Mark mbed-os-tests-mbed_hal-common_tickers/time_cpu_cycles not in-lined.
Otherwise, time_cpu_cycles may be in-lined somewhere and takes different
CPU cycles on invocation.

ccli8 committed:
[NANO130] Fix CLK_Idle incorrectly enters into deep sleep mode

This can happen with CLK_PowerDown() called first and then CLK_Idle() called.

ccli8 committed:
[NANO130] Adjust static/dynamic memory allocation for IAR toolchain to pass Greentea test

ccli8 commented:
### Greentea test (NUC472/ARM)

```
+------------------------+--------------------+----------------------------------------------------+--------+--------------------+-------------+
| target                 | platform_name      | test suite                                         | result | elapsed_time (sec) | copy_method |
+------------------------+--------------------+----------------------------------------------------+--------+--------------------+-------------+
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_drivers-lp_ticker               | OK     | 18.63              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_drivers-lp_timeout              | OK     | 56.3               | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_drivers-lp_timer                | OK     | 17.44              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_drivers-rtc                     | OK     | 27.3               | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_drivers-sleep_lock              | OK     | 12.95              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_drivers-ticker                  | OK     | 38.67              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_drivers-timeout                 | OK     | 54.97              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_drivers-timer                   | OK     | 20.03              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_drivers-timerevent              | OK     | 14.31              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-common_tickers              | FAIL   | 30.55              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-common_tickers_freq         | OK     | 34.1               | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-lp_ticker                   | OK     | 12.93              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-rtc_time                    | OK     | 14.24              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-rtc_time_conv               | OK     | 34.02              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-sleep_manager               | OK     | 12.04              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-sleep_manager_racecondition | OK     | 25.1               | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-ticker                      | OK     | 24.43              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-us_ticker                   | OK     | 13.03              | default     |
+------------------------+--------------------+----------------------------------------------------+--------+--------------------+-------------+
mbedgt: test suite results: 1 FAIL / 17 OK
```

ccli8 commented:
### Greentea test (M453/ARM)

```
+----------------------+------------------+----------------------------------------------------+--------+--------------------+-------------+
| target               | platform_name    | test suite                                         | result | elapsed_time (sec) | copy_method |
+----------------------+------------------+----------------------------------------------------+--------+--------------------+-------------+
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_drivers-lp_ticker               | OK     | 19.06              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_drivers-lp_timeout              | OK     | 57.99              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_drivers-lp_timer                | OK     | 18.45              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_drivers-rtc                     | OK     | 28.07              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_drivers-sleep_lock              | OK     | 13.01              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_drivers-ticker                  | OK     | 40.77              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_drivers-timeout                 | OK     | 54.29              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_drivers-timer                   | OK     | 20.76              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_drivers-timerevent              | OK     | 14.88              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_hal-common_tickers              | FAIL   | 30.87              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_hal-common_tickers_freq         | OK     | 34.46              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_hal-lp_ticker                   | OK     | 13.28              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_hal-rtc_time                    | OK     | 14.51              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_hal-rtc_time_conv               | OK     | 34.63              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_hal-sleep_manager               | OK     | 12.37              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_hal-sleep_manager_racecondition | OK     | 25.82              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_hal-ticker                      | OK     | 24.51              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_hal-us_ticker                   | OK     | 12.85              | default     |
+----------------------+------------------+----------------------------------------------------+--------+--------------------+-------------+
mbedgt: test suite results: 1 FAIL / 17 OK
```

ccli8 commented:
### Greentea test (M487/ARM)

```
+----------------------+------------------+----------------------------------------------------+--------+--------------------+-------------+
| target               | platform_name    | test suite                                         | result | elapsed_time (sec) | copy_method |
+----------------------+------------------+----------------------------------------------------+--------+--------------------+-------------+
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_drivers-lp_ticker               | OK     | 18.81              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_drivers-lp_timeout              | OK     | 58.08              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_drivers-lp_timer                | OK     | 18.13              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_drivers-rtc                     | OK     | 27.24              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_drivers-sleep_lock              | OK     | 12.23              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_drivers-ticker                  | OK     | 41.04              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_drivers-timeout                 | OK     | 54.71              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_drivers-timer                   | OK     | 20.48              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_drivers-timerevent              | OK     | 14.37              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_hal-common_tickers              | FAIL   | 30.26              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_hal-common_tickers_freq         | OK     | 35.32              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_hal-lp_ticker                   | OK     | 13.26              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_hal-rtc_time                    | OK     | 14.07              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_hal-rtc_time_conv               | OK     | 35.71              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_hal-sleep_manager               | OK     | 12.04              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_hal-sleep_manager_racecondition | OK     | 25.76              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_hal-ticker                      | OK     | 24.66              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_hal-us_ticker                   | OK     | 12.45              | default     |
+----------------------+------------------+----------------------------------------------------+--------+--------------------+-------------+
mbedgt: test suite results: 1 FAIL / 17 OK
```

ccli8 commented:
### Greentea test (NANO130/ARM)

```
+-------------------------+---------------------+----------------------------------------------------+--------+--------------------+-------------+
| target                  | platform_name       | test suite                                         | result | elapsed_time (sec) | copy_method |
+-------------------------+---------------------+----------------------------------------------------+--------+--------------------+-------------+
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_drivers-lp_ticker               | OK     | 18.92              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_drivers-lp_timeout              | OK     | 56.41              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_drivers-lp_timer                | OK     | 17.21              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_drivers-rtc                     | OK     | 28.0               | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_drivers-sleep_lock              | OK     | 13.21              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_drivers-ticker                  | OK     | 41.68              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_drivers-timeout                 | OK     | 52.78              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_drivers-timer                   | OK     | 20.09              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_drivers-timerevent              | OK     | 14.56              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_hal-common_tickers              | FAIL   | 30.59              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_hal-common_tickers_freq         | OK     | 35.46              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_hal-lp_ticker                   | OK     | 12.96              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_hal-rtc_time                    | OK     | 14.37              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_hal-rtc_time_conv               | OK     | 33.49              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_hal-sleep_manager               | OK     | 12.45              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_hal-sleep_manager_racecondition | OK     | 24.99              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_hal-ticker                      | OK     | 29.28              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_hal-us_ticker                   | OK     | 12.87              | default     |
+-------------------------+---------------------+----------------------------------------------------+--------+--------------------+-------------+
mbedgt: test suite results: 1 FAIL / 17 OK
```

ccli8 commented:
On all Nuvoton targets, `mbed-os-tests-mbed_hal-common_tickers`/`lp ticker speed test` is failed. This is H/W limitation. When Timer engine is clocked by low-power clock source **LXT**/**LIRC**, 3 engine clock delay is needed to make sure write access to it takes effect.

<pre>
void lp_ticker_disable_interrupt(void)
{
    TIMER_DisableInt((TIMER_T *) NU_MODBASE(TIMER_MODINIT.modname));
    <b>wait_us((NU_US_PER_SEC / NU_TMRCLK_PER_SEC) * 3);</b>
}
</pre>

```
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-common_tickers              | Microsecond ticker fire interrupt                               | 1      | 0      | OK     | 0.06               |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-common_tickers              | Microsecond ticker increment test                               | 1      | 0      | OK     | 0.09               |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-common_tickers              | Microsecond ticker info test                                    | 1      | 0      | OK     | 0.06               |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-common_tickers              | Microsecond ticker init is safe to call repeatedly              | 1      | 0      | OK     | 0.07               |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-common_tickers              | Microsecond ticker interrupt test                               | 1      | 0      | OK     | 0.06               |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-common_tickers              | Microsecond ticker overflow test                                | 1      | 0      | OK     | 12.88              |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-common_tickers              | Microsecond ticker past interrupt test                          | 1      | 0      | OK     | 0.08               |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-common_tickers              | Microsecond ticker reschedule test                              | 1      | 0      | OK     | 0.06               |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-common_tickers              | Microsecond ticker speed test                                   | 1      | 0      | OK     | 0.06               |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-common_tickers              | lp ticker fire interrupt                                        | 1      | 0      | OK     | 0.05               |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-common_tickers              | lp ticker increment test                                        | 1      | 0      | OK     | 0.06               |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-common_tickers              | lp ticker info test                                             | 1      | 0      | OK     | 0.04               |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-common_tickers              | lp ticker init is safe to call repeatedly                       | 1      | 0      | OK     | 0.08               |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-common_tickers              | lp ticker interrupt test                                        | 1      | 0      | OK     | 0.09               |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-common_tickers              | lp ticker overflow test                                         | 1      | 0      | OK     | 0.07               |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-common_tickers              | lp ticker past interrupt test                                   | 1      | 0      | OK     | 0.07               |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-common_tickers              | lp ticker reschedule test                                       | 1      | 0      | OK     | 0.1                |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-common_tickers              | lp ticker speed test                                            | 0      | 1      | FAIL   | 0.17               |
```

0xc0170 commented:
Same as for RTC branch  (my latest comment there) , this will be redirected to master once feature branch is integrated

ccli8 commented:
@0xc0170 I've done rebase onto master and change merge target branch to master.

cmonr commented:
@bulislaw Another HAL PR to check.

bulislaw commented:
@ccli8 can you use this mechanism to get that test to pass? https://github.com/ARMmbed/mbed-os/pull/6536

ccli8 commented:
@bulislaw #6536 (**LPTICKER_DELAY_TICKS**) doesn't help. Nuvoton's situation gets worse. In Nuvoton's `lp_ticker`, all Timer register writes need delay. #6536 just fixes `lp_ticker_set_interrupt` with one delay. Nuvoton's `lp_ticker_set_interrupt` needs two delays (two Timer register writes). Besides, other APIs (`lp_ticker_disable_interrupt`, `lp_ticker_clear_interrupt`, etc.) also need at least one delay (one Timer register write).

bulislaw commented:
@c1728p9 how suitable would be to extend what we have to cover all cases?

c1728p9 commented:
@bulislaw its hard to tell without knowing the exact use case. @ccli8 why does `lp_ticker_disable_interrupt` and `lp_ticker_clear_interrupt` require a delay? NVIC access should be immediate.

ccli8 committed:
[Nuvoton] Reduce blocking code in lp_ticker

1. Introduce S/W interrupt enable/disable to reduce calls to TIMER_EnableInt/TIMER_DisableInt.
2. Allow dummy interrupt because clear interrupt flag is not synchronized.
3. Enable LPTICKER_DELAY_TICKS to make lp_ticker_set_interrupt non-blocking.

ccli8 committed:
[Nuvoton] Synchronize lp_ticker code to us_ticker

This is to make us_ticker/lp_ticker code consistent.

ccli8 commented:
> why does `lp_ticker_disable_interrupt` and `lp_ticker_clear_interrupt` require a delay? NVIC access should be immediate.

These APIs have write access to Timer registers. Anyway, I will seek another means to make these APIs non-blocking.

Besides, I have questions about these ticker APIs:
1. In `lp_ticker_disable_interrupt`, I shall disable Timer interrupt or NVIC interrupt (`NVIC_DisableIRQ`), or both?
1. In `lp_ticker_clear_interrupt`, I shall clear Timer interrupt flag or also NVIC interrupt flag (`NVIC_ClearPendingIRQ`)?

@bulislaw @c1728p9

c1728p9 commented:
> In lp_ticker_disable_interrupt, I shall disable Timer interrupt or NVIC interrupt (NVIC_DisableIRQ), or both?

This is up to you @ccli8. Whichever is more straight forward for your to implement. The only requirement is that once `lp_ticker_disable_interrupt` is called you don't get interrupts. In this case it sounds like NVIC_DisableIRQ may be easier as it does not require a delay.

> In lp_ticker_clear_interrupt, I shall clear Timer interrupt flag or also NVIC interrupt flag (NVIC_ClearPendingIRQ)?

The NVIC interrupt should get cleared automatically when you return from the given interrupt handler, so you should only need to clear the timer flag.

ccli8 commented:
@c1728p9 I fix the delay issue in https://github.com/ARMmbed/mbed-os/pull/7029/commits/4668c8fbd36ba066209f66b52c7f967062a54d08.

ccli8 commented:
### NUC472/ARM

```
+------------------------+--------------------+----------------------------------------------------+--------+--------------------+-------------+
| target                 | platform_name      | test suite                                         | result | elapsed_time (sec) | copy_method |
+------------------------+--------------------+----------------------------------------------------+--------+--------------------+-------------+
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_drivers-lp_ticker               | OK     | 18.75              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_drivers-lp_timeout              | OK     | 57.41              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_drivers-lp_timer                | OK     | 17.53              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_drivers-rtc                     | OK     | 27.3               | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_drivers-sleep_lock              | OK     | 12.04              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_drivers-ticker                  | OK     | 39.98              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_drivers-timeout                 | OK     | 55.3               | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_drivers-timer                   | OK     | 19.55              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_drivers-timerevent              | OK     | 13.93              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-common_tickers              | OK     | 29.75              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-common_tickers_freq         | OK     | 34.35              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-lp_ticker                   | OK     | 12.78              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-rtc                         | OK     | 55.26              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-rtc_reset                   | OK     | 17.63              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-rtc_time                    | OK     | 13.91              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-rtc_time_conv               | OK     | 34.77              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-sleep_manager               | OK     | 11.91              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-sleep_manager_racecondition | OK     | 24.8               | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-ticker                      | OK     | 24.21              | default     |
| NUMAKER_PFM_NUC472-ARM | NUMAKER_PFM_NUC472 | mbed-os-tests-mbed_hal-us_ticker                   | OK     | 13.26              | default     |
+------------------------+--------------------+----------------------------------------------------+--------+--------------------+-------------+
mbedgt: test suite results: 20 OK
```

ccli8 commented:
### M453/ARM

```
+----------------------+------------------+----------------------------------------------------+--------+--------------------+-------------+
| target               | platform_name    | test suite                                         | result | elapsed_time (sec) | copy_method |
+----------------------+------------------+----------------------------------------------------+--------+--------------------+-------------+
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_drivers-lp_ticker               | OK     | 19.13              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_drivers-lp_timeout              | OK     | 57.16              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_drivers-lp_timer                | OK     | 18.0               | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_drivers-rtc                     | OK     | 29.16              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_drivers-sleep_lock              | OK     | 12.75              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_drivers-ticker                  | OK     | 40.72              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_drivers-timeout                 | OK     | 54.34              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_drivers-timer                   | OK     | 20.39              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_drivers-timerevent              | OK     | 14.75              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_hal-common_tickers              | OK     | 30.62              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_hal-common_tickers_freq         | OK     | 36.11              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_hal-lp_ticker                   | OK     | 13.31              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_hal-rtc                         | OK     | 57.61              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_hal-rtc_reset                   | OK     | 18.22              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_hal-rtc_time                    | OK     | 14.76              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_hal-rtc_time_conv               | OK     | 35.76              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_hal-sleep_manager               | OK     | 12.78              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_hal-sleep_manager_racecondition | OK     | 25.63              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_hal-ticker                      | OK     | 24.88              | default     |
| NUMAKER_PFM_M453-ARM | NUMAKER_PFM_M453 | mbed-os-tests-mbed_hal-us_ticker                   | OK     | 13.54              | default     |
+----------------------+------------------+----------------------------------------------------+--------+--------------------+-------------+
mbedgt: test suite results: 20 OK
```

ccli8 commented:
### M487/ARM

```
+----------------------+------------------+----------------------------------------------------+--------+--------------------+-------------+
| target               | platform_name    | test suite                                         | result | elapsed_time (sec) | copy_method |
+----------------------+------------------+----------------------------------------------------+--------+--------------------+-------------+
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_drivers-lp_ticker               | OK     | 18.75              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_drivers-lp_timeout              | OK     | 56.22              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_drivers-lp_timer                | OK     | 17.27              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_drivers-rtc                     | OK     | 27.33              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_drivers-sleep_lock              | OK     | 12.34              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_drivers-ticker                  | OK     | 39.9               | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_drivers-timeout                 | OK     | 54.91              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_drivers-timer                   | OK     | 19.66              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_drivers-timerevent              | OK     | 14.04              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_hal-common_tickers              | OK     | 30.14              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_hal-common_tickers_freq         | OK     | 35.4               | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_hal-lp_ticker                   | OK     | 12.53              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_hal-rtc                         | OK     | 55.55              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_hal-rtc_reset                   | OK     | 17.42              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_hal-rtc_time                    | OK     | 14.18              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_hal-rtc_time_conv               | OK     | 34.91              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_hal-sleep_manager               | OK     | 12.14              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_hal-sleep_manager_racecondition | OK     | 25.04              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_hal-ticker                      | OK     | 24.12              | default     |
| NUMAKER_PFM_M487-ARM | NUMAKER_PFM_M487 | mbed-os-tests-mbed_hal-us_ticker                   | OK     | 11.95              | default     |
+----------------------+------------------+----------------------------------------------------+--------+--------------------+-------------+
mbedgt: test suite results: 20 OK
```

ccli8 commented:
### NANO130/ARM

```
+-------------------------+---------------------+----------------------------------------------------+--------+--------------------+-------------+
| target                  | platform_name       | test suite                                         | result | elapsed_time (sec) | copy_method |
+-------------------------+---------------------+----------------------------------------------------+--------+--------------------+-------------+
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_drivers-lp_ticker               | OK     | 18.33              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_drivers-lp_timeout              | OK     | 56.24              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_drivers-lp_timer                | OK     | 17.27              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_drivers-rtc                     | OK     | 27.53              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_drivers-sleep_lock              | OK     | 12.29              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_drivers-ticker                  | OK     | 41.09              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_drivers-timeout                 | OK     | 53.71              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_drivers-timer                   | OK     | 19.66              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_drivers-timerevent              | OK     | 13.99              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_hal-common_tickers              | OK     | 30.11              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_hal-common_tickers_freq         | OK     | 32.84              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_hal-lp_ticker                   | OK     | 12.67              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_hal-rtc                         | OK     | 55.38              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_hal-rtc_reset                   | OK     | 17.29              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_hal-rtc_time                    | OK     | 13.98              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_hal-rtc_time_conv               | OK     | 34.35              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_hal-sleep_manager               | OK     | 12.15              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_hal-sleep_manager_racecondition | OK     | 24.98              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_hal-ticker                      | OK     | 29.17              | default     |
| NUMAKER_PFM_NANO130-ARM | NUMAKER_PFM_NANO130 | mbed-os-tests-mbed_hal-us_ticker                   | OK     | 12.03              | default     |
+-------------------------+---------------------+----------------------------------------------------+--------+--------------------+-------------+
mbedgt: test suite results: 20 OK
```

ccli8 commented:
@bulislaw @c1728p9 
Any update?

0xc0170 commented:
Needs @ARMmbed/mbed-os-hal  approvals!

c1728p9 commented on targets/TARGET_NUVOTON/TARGET_M480/lp_ticker.c:
```diff
@@ -163,14 +222,15 @@ const ticker_info_t* lp_ticker_get_info()
+    /* NOTE: We need to clear interrupt flag earlier to reduce possibility of dummy interrupt.
+     * This is because "clear interrupt flag" needs delay which isn't added here to avoid
+     * blocking in ISR code. */
+    lp_ticker_clear_interrupt();
```
does this clear take effect immediately, or is the interrupt stuck pending for a period of time (causing tmr1_vec to get called multiple times)?

ccli8 replied:
@c1728p9 Yes. This clear doesn't take effect immediately. If the following `lp_ticker_irq_handler` takes enough time, `tmr1_vec` will be called only once. If `lp_ticker_irq_handler` takes very short time or even doesn't get called, `tmr1_vec` would get called multiple times until the first clear takes effect.

c1728p9 replied:
Having tmr1_vec getting stuck pending is a problem as well because lower priority interrupts or threads won't get a chance to run. As an alternative fix, would it be possible to disable the NVIC interrupt in place of setting NU_TICKER_INIT_INTR_DIS and clear+enable in place of NU_TICKER_INIT_INTR_EN? That way the interrupt won't fire repeatedly.

ccli8 replied:
@c1728p9 That doesn't help. As `lp_ticker_set_interrupt` (`NVIC_EnableIRQ`) or `lp_ticker_fire_interrupt` (`NVIC_EnableIRQ`) are called in `lp_ticker_irq_handler`, `tmr1_vec` (`NVIC_DisableIRQ`) is still getting stuck. We will be in a loop of `tmr1_vec` > `lp_ticker_fire_interrupt`/`lp_ticker_fire_interrupt` > `tmr1_vec` > `lp_ticker_fire_interrupt`/`lp_ticker_fire_interrupt` until clear flag takes effect.

ccli8 replied:
@c1728p9 Correct my comment above. Your idea would be workable. We won't get stuck in a loop of `tmr1_vec` (clear interrupt flag and disable interrupt at the start) and `lp_ticker_fire_interrupt`/`lp_ticker_set_interrupt` (called inside `lp_ticker_irq_handler` and so interrupt re-enabled) before the above "clear interrupt flag" takes effect:
1. When `lp_ticker_fire_interrupt` is called, it means there is past event and so the interrupt with uncompleted "clear interrupt flag" gets not dummy.
1. With `LPTICKER_DELAY_TICKS` enabled, it is `lp_ticker_set_interrupt_wrapper` rather than `lp_ticker_set_interrupt` that gets straight called by `lp_ticker_irq_handler`. `lp_ticker_set_interrupt_wrapper` guarantees that `lp_ticker_set_interrupt` won't get re-called in `LPTICKER_DELAY_TICKS` ticks which is just enough for  "clear interrupt flag" to take effect.

This idea is implemented in https://github.com/ARMmbed/mbed-os/pull/7029/commits/26e24d604ea43a0d53fc07b96a32ef5e80da9f9b.

ccli8 replied:
@c1728p9 Any update?

ccli8 committed:
[Nuvoton] Fix trap in lp_ticker ISR with non-blocking "clear interrupt flag"

ccli8 committed:
[Nuvoton] Synchronize lp_ticker code to us_ticker

This is to make us_ticker/lp_ticker code consistent.

bulislaw commented:
@c1728p9 please re-review

ccli8 commented:
Do rebase on master to fix conflict

c1728p9 commented:
Hi @ccli8 I took an in-depth look into this on the NUMAKER_PFM_M453. I was unable to reproduce the problems you describe with requiring a delay to clear the interrupt. I used GPIO to record the timing of `tmr1_vec`, `lp_ticker_clear_interrupt` and `lp_ticker_set_interrupt`. I found that even if I removed `NVIC_DisableIRQ` from `tmr1_vec` the interrupt would not fire repeatedly. Further, writes to `timer_base->CMP` and `timer_base->INTSTS` seemed to effect immediately. Even if I waited in a loop for the operation to take effect (reading CMP or INTSTS until it matched expected) neither took more than 5us. This is much shorter than the ~30us of a single low power clock cycle.

A screenshots of the timing can be seen below. None of the functions take more than 10us to execute. Test code used for this screenshot and for the testing above can be found on the [lp_timer_test_code branch of my mbed-os repo](https://github.com/c1728p9/mbed-os/tree/lp_timer_test_code).

![image](https://user-images.githubusercontent.com/2916772/41986977-7997b46a-79fd-11e8-8094-538508e67e15.png)

Further I looked through the Reference Manual but didn't see any information about synchronization delays when using the TMR module. Could this already be handled by hardware?

Summary - @ccli8 would it be possible to remove the special handling needed for `lp_ticker_disable_interrupt`? If not could you point me to documentation or example code showing why or demonstrating the failures you are seeing?

ccli8 committed:
[Nuvoton] Remove special handling for dummy interrupt in lp_ticker

It is because dummy interrupt is very rare or pending time caused by it
is very short.

ccli8 committed:
[Nuvoton] Synchronize lp_ticker code to us_ticker

This is to make us_ticker/lp_ticker code consistent.

ccli8 commented:
@c1728p9 Many thanks for your in-depth look. Re-checking with our designer, I have to fix my point. For **LXT**/**LIRC**-clocked Timer (**lp_ticker**), there are synchronization issues with register control. Before that, I have to clarify two clock domains:
- PCLK: typical HCLK/2
- ECLK (engine clock): LXT/LIRC (30us/100us) for Timer used to implement lp_ticker

1. For write `CMP` to take effect, it takes typical 1PCLK + 2ECLK. Read-check is not enough because it just checks PCLK domain and doesn't check into ECLK domain.
 1. For **clear interrupt flag** to take effect, it takes typical 2PCLK. It is very rare that we would meet dummy interrupt and get stuck in ISR until 'clear interrupt flag' takes effect. The issue is ignorable because the pending time is very short (at most 1 dummy interrupt). This also explains why you didn't meet dummy interrupt (and I also didn't). Needing no special handling for it, I remove `lp_ticker_disable_interrupt` in ISR.

c1728p9 commented on targets/TARGET_NUVOTON/TARGET_M451/lp_ticker.c:
```diff
@@ -129,27 +181,39 @@ void lp_ticker_set_interrupt(timestamp_t timestamp)
-    TIMER_ClearIntFlag((TIMER_T *) NU_MODBASE(TIMER_MODINIT.modname));
-    wait_us((NU_US_PER_SEC / NU_TMRCLK_PER_SEC) * 3);
+    /* To avoid sync issue, we clear TIF/TWKF simultaneously rather than call separate 
+     * driver API:
```
Since this is on the PCLK domain does writing these separately cause issues? If not you could revert this change.

ccli8 replied:
@c1728p9 `TIMER_ClearIntFlag`/`TIMER_ClearWakeupFlag` are so close. It is safer to clear these interrupt flags altogether.

c1728p9 commented on targets/TARGET_NUVOTON/TARGET_M451/lp_ticker.c:
```diff
@@ -101,6 +126,33 @@ void lp_ticker_init(void)
+
+    /* Disable wakeup */
+    TIMER_DisableWakeup(timer_base);
+    wait_us((NU_US_PER_SEC / NU_TMRCLK_PER_SEC) * 3);
```
Is this `wait_us` needed? Same question for the `wait_us` below.

ccli8 replied:
@c1728p9 All these `wait_us` are necessary.

c1728p9 commented on targets/TARGET_NUVOTON/TARGET_M451/lp_ticker.c:
```diff
@@ -42,17 +42,42 @@ static const struct nu_modinit_s timer1_modinit = {TIMER_1, TMR1_MODULE, CLK_CLK
+ *    It is very rare that we would meet dummy interrupt and get stuck in ISR until
+ *    'clear interrupt flag' takes effect. The issue is ignorable because the pending
+ *    time is very short (at most 1 dummy interrupt). We won't take special handling for it.
+ */
```
I like this comment 👍

ccli8 replied:
@c1728p9 Re-thanks for your detailed review.

cmonr commented:
/morph build

cmonr commented on TESTS/mbed_hal/common_tickers/main.cpp:
```diff
@@ -52,6 +52,7 @@ unsigned int ticker_overflow_delta;
 
 /* Auxiliary function to count ticker ticks elapsed during execution of N cycles of empty while loop.
  * Parameter <step> is used to disable compiler optimisation. */
+MBED_NOINLINE
```
Why was this inserted?

cmonr replied:
Never mind. Commit b18f690b39d9d1ad825375665b6281fc15ad4a74 explained it perfectly.

cmonr commented:
/morph build
